unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -125,30 +129,37 @@ public UTF8String substring(final int start, final int until) {
   }
 
   public boolean contains(final UTF8String substring) {
+    if (substring == null) return false;
     final byte[] b = substring.getBytes();
     if (b.length == 0) {
       return true;
     }
 
     for (int i = 0; i <= bytes.length - b.length; i++) {
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(b, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final byte[] prefix, int offset) {
```

Comment:
This is adopted from the original java String API. But yes it's true that ```offset``` depends on the bytes of the string and not on characters. Shall I rename it to ```match```? I could add a comment that ```offset`` depends on the bytes representation and not on characters